File: clients/typescript/store_client.test.ts
import { test, before, after } from "node:test";
import assert from "node:assert";
import { GenericContainer, StartedTestContainer } from "testcontainers";
import { SharedStoreClient } from "./store_client";
import { randomUUID } from "node:crypto";

let container: StartedTestContainer;
let redisHost: string;
let redisPort: number;

before(async () => {
  // Start the container
  container = await new GenericContainer("redis:7.0-alpine")
    .withExposedPorts(6379)
    .start();
  
  redisHost = container.getHost();
  redisPort = container.getMappedPort(6379);
});

after(async () => {
  if (container) {
    await container.stop();
  }
});

function createClient(): SharedStoreClient {
  const namespace = `test-ts-${randomUUID().slice(0, 8)}`;
  return new SharedStoreClient({
    host: redisHost,
    port: redisPort,
    namespace
  });
}

test("TypeScript Client - put, get and secondary index lookups", async () => {
  const client = createClient();
  const primary = "comp-123";
  const secondary = "acc-456";
  const payload = { companyName: "Antigravity TS", count: 42 };

  // Test put
  const putOk = await client.put(primary, secondary, payload, 10);
  assert.strictEqual(putOk, true);

  // Test get by primary
  const item = await client.get(primary);
  assert.ok(item);
  assert.strictEqual(item.primary_val, primary);
  assert.strictEqual(item.secondary_val, secondary);
  assert.deepStrictEqual(item.payload, payload);

  // Test get by secondary index
  const itemBySec = await client.getBySecondary(secondary);
  assert.ok(itemBySec);
  assert.strictEqual(itemBySec.primary_val, primary);
  assert.deepStrictEqual(itemBySec.payload, payload);

  await client.close();
});

test("TypeScript Client - delete removes secondary index", async () => {
  const client = createClient();
  const primary = "comp-789";
  const secondary = "acc-012";
  const payload = { deleteMe: true };

  await client.put(primary, secondary, payload, 10);
  const delOk = await client.delete(primary);
  assert.strictEqual(delOk, true);

  const item = await client.get(primary);
  assert.strictEqual(item, null);

  const itemBySec = await client.getBySecondary(secondary);
  assert.strictEqual(itemBySec, null);

  await client.close();
});

test("TypeScript Client - TTL expiration", async () => {
  const client = createClient();
  const primary = "comp-ttl";
  const secondary = "acc-ttl";
  const payload = { temp: true };

  await client.put(primary, secondary, payload, 1);
  
  // Wait 1.5 seconds for TTL expiration
  await new Promise((resolve) => setTimeout(resolve, 1500));

  const item = await client.get(primary);
  assert.strictEqual(item, null);

  const itemBySec = await client.getBySecondary(secondary);
  assert.strictEqual(itemBySec, null);

  await client.close();
});

File: clients/elixir/test/store_client_test.exs
defmodule SharedStoreClientTest do
  use ExUnit.Case, async: true
  import Testcontainers.ExUnit

  # Automatically start an ephemeral Redis container for this test suite
  container(
    :redis,
    Testcontainers.RedisContainer.new()
    |> Testcontainers.RedisContainer.with_image("redis:7.0-alpine")
  )

  setup %{redis: redis} do
    host = Testcontainers.get_host()
    port = Testcontainers.Container.mapped_port(redis, 6379)
    # Generate unique namespace to support parallel test execution
    namespace = "test-ex-#{:erlang.unique_integer([:positive])}"

    {:ok, client} = SharedStoreClient.connect(host: host, port: port, namespace: namespace)

    on_exit(fn ->
      SharedStoreClient.disconnect(client)
    end)

    {:ok, client: client}
  end

  test "put, get and secondary index queries", %{client: client} do
    primary = "company_999"
    secondary = "account_888"
    payload = %{"name" => "Antigravity Elixir", "tags" => ["tech", "cloud"]}

    # Put data
    assert :ok == SharedStoreClient.put(client, primary, secondary, payload, 10)

    # Get data by primary key
    assert {:ok, retrieved} = SharedStoreClient.get(client, primary)
    assert retrieved["primary_val"] == primary
    assert retrieved["secondary_val"] == secondary
    assert retrieved["payload"] == payload

    # Get data by secondary index
    assert {:ok, retrieved_by_sec} = SharedStoreClient.get_by_secondary(client, secondary)
    assert retrieved_by_sec["primary_val"] == primary
    assert retrieved_by_sec["payload"] == payload
  end

  test "delete removes index and data", %{client: client} do
    primary = "company_delete"
    secondary = "account_delete"
    payload = %{"to_delete" => true}

    assert :ok == SharedStoreClient.put(client, primary, secondary, payload, 10)
    assert :ok == SharedStoreClient.delete(client, primary)

    assert {:ok, nil} == SharedStoreClient.get(client, primary)
    assert {:ok, nil} == SharedStoreClient.get_by_secondary(client, secondary)
  end

  test "TTL expiration removes data", %{client: client} do
    primary = "company_ttl"
    secondary = "account_ttl"
    payload = %{"temporary" => true}

    # Write with 1 second TTL
    assert :ok == SharedStoreClient.put(client, primary, secondary, payload, 1)

    # Wait 1.5 seconds for TTL expiration
    Process.sleep(1500)

    assert {:ok, nil} == SharedStoreClient.get(client, primary)
    assert {:ok, nil} == SharedStoreClient.get_by_secondary(client, secondary)
  end
end
